articles/Frontend Performance Optimization/README.md

```diff
@@ -0,0 +1,519 @@
+- Long-running loops exceeding 50 milliseconds.
+
+```ts
+const observer = new PerformanceObserver((list) => {
+  list.getEntries().forEach((entry) => {
+    console.log(entry);
+  });
+});
+
+observer.observe({ type: "longtask", buffered: true });
+```
+
+![long task object](long-task.png)
+There are several metrics within this context:
+
+- **`duration`**: Represents the duration of the task, i.e., the time elapsed from start to finish.
+- **`TaskAttributionTiming`**: This is an object associated with Long Tasks, used to track and attribute the execution of long tasks. This object may contain detailed information about the long task, such as its source, triggering events, etc. Through this object, developers can gain a better understanding of the context and reasons for long tasks, facilitating performance optimization and debugging.
+
+Because long tasks significantly impact user experience, they are separately highlighted even though they are not part of Web Vitals.
+
+## Optimization Measures
+
+### Code splitting and lazy loading
+
+Lazy loading and code splitting are both strategies used to optimize frontend performance, but they have different goals and applications.
+
+1. **Code Splitting**:
+
+   **Goal**: The main aim of code splitting is to reduce the initial JavaScript file size required during the loading phase, improving the initial page loading speed. It breaks down the application code into multiple chunks, often based on **routes** or **features**, allowing for **on-demand loading**.
+
+2. **Lazy Loading**:
+
+   **Goal**: Lazy loading is a strategy that allows you to **defer the loading** of certain components, resources, or functionalities, rather than loading them during the initial page load. This helps reduce the resource burden during the initial load, improving page speed.
+
+In the context of React, these concepts are often applied by combining code splitting with lazy loading.
+
+```ts
+const DownloadFile = lazy(() => import("./page/OperateFile/OperateFile"));
+const TimeSelect = lazy(() => import("./page/TimeSelect/TimeSelect"));
+
+export const router: Router[] = [
+  {
+    path: "/",
+    element: <App />,
+    name: "Home",
+  },
+  {
+    path: "/download-file",
+    element: <DownloadFile />,
+    name: "Download File",
+  },
+  {
+    path: "/time-select",
+    element: <TimeSelect />,
+    name: "Time Select",
+  },
+];
+```
+
+### Http cache
+
+Usually, we configure appropriate directives based on the update frequency of resources to retrieve them from the cache. This helps reduce request frequency and improve loading efficiency. This involves specific configurations for responses and requests. The specific configuration directives may vary depending on the circumstances. For detailed information, please click [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control#cache_directives).
+
+1. For HTML files with a high update frequency, the directive settings commonly used are:
+
+```ts
```

Comment:
If the code is in a list, it is best to indent the code to ensure that the code is aligned with the list item.

This is also the case elsewhere in the article.